File: app/src/main/java/com/example/jetpackdemo/android_view_model_04/MyViewModel.java
package com.example.jetpackdemo.android_view_model_04;

import android.app.Application;
import android.content.Context;
import android.content.SharedPreferences;

import androidx.annotation.NonNull;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.SavedStateHandle;
import androidx.lifecycle.ViewModel;

import com.example.jetpackdemo.R;

public class MyViewModel extends AndroidViewModel {

    String shpName = getApplication().getResources().getString(R.string.shp_name);
    String key =getApplication().getResources().getString(R.string.data_key);
    SavedStateHandle handle;

    public MyViewModel(@NonNull Application application,SavedStateHandle handle) {
        super(application);
        this.handle =  handle;

        if (!handle.contains(key)){
            load();
        }
    }

    public LiveData<Integer> getNumber(){
        return handle.getLiveData(key);
    }


    private void load() {
        SharedPreferences shp = getApplication().getSharedPreferences(shpName,Context.MODE_PRIVATE);
        int x = shp.getInt(key,0);
        handle.set(key,x);
    }


    void save(){
        SharedPreferences shp = getApplication().getSharedPreferences(shpName,Context.MODE_PRIVATE);
        SharedPreferences.Editor editor = shp.edit() ;
        editor.putInt(key,getNumber().getValue());
        editor.apply();
    }

    void add(int x ){
        handle.set(key,getNumber().getValue() + x);
        save();
    }

}

File: app/src/main/java/com/example/jetpackdemo/day15_new_livedata/MyViewModel.kt
package com.example.jetpackdemo.day15_new_livedata

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.SavedStateHandle

class MyViewModel(private var savedStateHandle: SavedStateHandle,application: Application) : AndroidViewModel(application) {

    private val _text  = MutableLiveData<Int>().also {
        if (!savedStateHandle.contains("number")){
            savedStateHandle.set("number",0)
        }
        it.value = savedStateHandle.get("number")
    }
     val text1 :LiveData<Int> get() = _text

    public fun addOne(){
        _text.value = text1.value?.plus(1)
        savedStateHandle.set("number",_text.value)
    }
}